From 7e650627b0390bb553316004c665a7a0e1651d78 Mon Sep 17 00:00:00 2001
From: Brian Bockelman <bbockelm@cse.unl.edu>
Date: Thu, 24 Apr 2014 09:25:37 -0500
Subject: [PATCH] Correctly handle reads of a partial header.  #45

If a partial header was read, we previously did not return to the
poll loop.  Instead, a read into the data buffer was attempted.  If
that read actually succeeded (that is, a packet arrived between the
two read calls), then the next attempt to read the header resulted
in garbage size/offsets being attempted.

The fix is to simply check for a partial header read and return to the
poll loop immediately if one occurred.
---
 src/XrdCl/XrdClXRootDMsgHandler.cc | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/XrdCl/XrdClXRootDMsgHandler.cc b/src/XrdCl/XrdClXRootDMsgHandler.cc
index 5eca68a..aa50564 100644
--- a/src/XrdCl/XrdClXRootDMsgHandler.cc
+++ b/src/XrdCl/XrdClXRootDMsgHandler.cc
@@ -797,6 +797,12 @@ namespace XrdCl
       //------------------------------------------------------------------------
       if( !st.IsOK() )
         return st;
+
+      //------------------------------------------------------------------------
+      // If we are not done reading the header, return back to the event loop.
+      //------------------------------------------------------------------------
+      if( st.IsOK() && st.code != suDone )
+        return st;
     }
 
     //--------------------------------------------------------------------------
-- 
1.9.1

